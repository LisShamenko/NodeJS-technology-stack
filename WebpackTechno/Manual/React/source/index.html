<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <title>Плагин 'html-webpack-plugin'.</title>
</head>

<body>
    <div id="root"></div>
</body>

</html>