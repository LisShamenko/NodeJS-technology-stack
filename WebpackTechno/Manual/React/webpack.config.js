// --------------- 2. React.

// Для транспиляции кода с компонентами React требуется добавить предустановку 
//      babel для React.
//      npm i @babel/core babel-loader @babel/preset-env @babel/preset-react --save-dev

// Файл настройки babel для использования предустановки React.
//      {
//          "presets": [
//              "@babel/preset-env", 
//              "@babel/preset-react"
//          ]
//      }

// Webpack сам добавляет ссылку на файл bundle.js в index.html.

const path = require('path');
const HtmlWebpackPlugin = require("html-webpack-plugin");

module.exports = {
    mode: 'development',
    entry: {
        index: path.resolve(__dirname, "source", "index.js")
    },
    output: {
        path: path.resolve(__dirname, "build"),
        filename: 'bundle.js'
    },
    plugins: [
        new HtmlWebpackPlugin({
            template: path.resolve(__dirname, "source", "index.html")
        })
    ],
    module: {
        rules: [
            {
                test: /\.css$/,
                use: ["style-loader", "css-loader"]
            },
            {
                test: /\.scss$/,
                use: ["style-loader", "css-loader", "sass-loader"]
            },
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        presets: [
                            '@babel/preset-env',
                            "@babel/preset-react"
                        ],
                        plugins: [
                            "@babel/plugin-proposal-class-properties",
                            "@babel/plugin-transform-runtime"
                        ]
                    }
                }
            }
        ]
    },
};