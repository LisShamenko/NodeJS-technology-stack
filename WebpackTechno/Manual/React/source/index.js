import React, { useState, Fragment } from "react";
import { render } from "react-dom";



// --------------- 2. CSS в React.

import "./style.css";
import "./style.scss";

// Способы применения CSS в React:
//      - CSS-in-CSS (CSS, Sass, Модули CSS, Модули CSS и Sass)
//      - CSS-in-JS (Styled Components, Emotion)
//      - Utility-First-CSS (Tailwind CSS)

import Basket_base from "./components/basket_base";
import Basket_file_css from "./components/basket_file_css";
import Basket_file_scss from "./components/basket_file_scss";
import Basket_module_css from "./components/basket_module_css";
import Basket_styled_components from "./components/basket_styled_components";
import Basket_inline_css from "./components/basket_inline_css";

// 
function App() {

    // 
    const [state, setState] = useState("CLICK ME");

    // 
    const [fruits, setFruits] = React.useState([
        { id: '1', name: 'Apple', isFavorite: false },
        { id: '2', name: 'Peach', isFavorite: true },
        { id: '3', name: 'Strawberry', isFavorite: false },
    ]);

    //
    function handleClick(item) {
        const newFruits = fruits.map((fruit) => {
            if (fruit.id === item.id) {
                return {
                    id: fruit.id,
                    name: fruit.name,
                    isFavorite: !fruit.isFavorite,
                };
            }
            else {
                return fruit;
            }
        });
        setFruits(newFruits);
    }

    // 
    return (
        <Fragment>
            <div>
                <h1 className="html">Плагин HTML.</h1>
                <h1 className="sass">Загрузчик SASS.</h1>
                <button onClick={() => setState('click')} >
                    {state}
                </button>
                <hr />
            </div>
            <div>
                <h3>No style.</h3>
                <Basket_base items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Файл CSS.</h3>
                <Basket_file_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Файл SCSS.</h3>
                <Basket_file_scss items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Модульный CSS.</h3>
                <Basket_module_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Библиотека styled-components.</h3>
                <Basket_styled_components items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Встроенные стили React.</h3>
                <Basket_inline_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
        </Fragment>
    );
}

// 
render(<App />, document.getElementById("root"));