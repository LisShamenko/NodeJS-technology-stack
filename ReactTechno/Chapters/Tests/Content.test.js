// Jasmine-style - стиль описания тестов при помощи describe и it.
//      https://jasmine.github.io/



import React from 'react';
import renderer from 'react-test-renderer';

// тестируемый компонент
import Content from './components/Content';

// Enzyme позволяет создавать пригодную для тестирования обертку компонента, при этом
//      компонент не встраивается и DOM не создается.
import { shallow } from 'enzyme';



// --------------- 10. Тестирование.

// При использовании методологии TDD при написании теста продумывается работа 
//      компонента, после чего реализуется сам компонент.

// describe делит тесты на отдельные звенья для тестирования разных частей кода
describe('component Content', () => {

    // тесты группируются при помощи describe
    describe('render', () => {

        // объявление теста, можно использовать it или test
        //      test('first test', () => { });

        // тесты должны быть простыми, особенно для крупных и сложных компонентов

        // 
        it('shallow', () => {

            // фиктивный объект
            const mockPost = {
                content: 'Контент.',
            };

            // Компоненты React должно быть простыми. Компонет следует разбивать 
            //      на части, если для его тестирования требуется множество тестов.

            // выполнить рендеринг компонента и вернуть обертку компонента
            const wrapper = shallow(<Content post={mockPost} />);

            // для выполнения утверждений испольузется встроенная функция expect,
            //      если утверждение не подтверждается, то тест терпит неудачу
            expect(wrapper.find('p').length).toBe(1);
            expect(wrapper.find('p.content').length).toBe(1);
            expect(wrapper.find('.content').text()).toBe(mockPost.content);
            expect(wrapper.find('p').text()).toBe(mockPost.content);
        });

        // 
        test('snapshot', () => {

            // 
            const mockPost = {
                content: 'Контент.',
            };

            // 
            const component = renderer.create(<Content post={mockPost} />);

            // Тестирование моментальных снимков - это процес, в котором тестируемый 
            //      визуальный вывод сравнивается с эталонным выводом. Jest будет 
            //      создавать JSON для тестов и хранить их в специальных каталогах, 
            //      которые должны быть добавлены в систему контроля.

            // 
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});