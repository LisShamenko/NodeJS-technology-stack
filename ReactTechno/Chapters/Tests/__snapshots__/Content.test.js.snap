// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Content render snapshot 1`] = `
<p
  className="content"
>
  Контент.
</p>
`;
